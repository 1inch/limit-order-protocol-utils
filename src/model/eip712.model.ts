export const ORDER_STRUCTURE = [
    { name: 'salt', type: 'uint256' },
    { name: 'maker', type: 'address' },
    { name: 'receiver', type: 'address' },
    { name: 'makerAsset', type: 'address' },
    { name: 'takerAsset', type: 'address' },
    { name: 'makingAmount', type: 'uint256' },
    { name: 'takingAmount', type: 'uint256' },
    { name: 'makerTraits', type: 'uint256' },
];

export interface EIP712TypedData {
    types: EIP712Types & { EIP712Domain: EIP712Parameter[] };
    domain: EIP712Object;
    message: EIP712Object;
    primaryType: string;
}

export interface EIP712Types {
    [key: string]: EIP712Parameter[];
}

export interface EIP712Parameter {
    name: string;
    type: string;
}

export declare type EIP712ObjectValue = string | number | EIP712Object;

export interface EIP712Object {
    [key: string]: EIP712ObjectValue;
}

export interface MessageTypes {
    [additionalProperties: string]: MessageTypeProperty[];
    EIP712Domain: MessageTypeProperty[];
}

export interface MessageTypeProperty {
    name: string;
    type: string;
}
