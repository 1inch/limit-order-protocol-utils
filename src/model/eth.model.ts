/**
 * Eth address 160bits
 */
export type Address = string;