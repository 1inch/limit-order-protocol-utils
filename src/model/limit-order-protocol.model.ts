import {LimitOrderPredicateCallData} from '../limit-order-predicate.builder';

export enum ChainId {
    etherumMainnet = 1,
    binanceMainnet = 56,
    polygonMainnet = 137,
    optimismMainnet = 10,
    arbitrumMainnet = 42161,
    gnosisMainnet = 100,
    avalancheMainnet = 43114,
    fantomMainnet = 250,
    auroraMainnet = 1313161554,
    klaytnMainnet = 8217,
    zkSyncEraMainnet = 324,
}

export type LimitOrderSignature = string;

export type LimitOrderHash = string;

// RFQOrderData.expiresInTimestamp | RFQOrderData.id
export type RFQOrderInfo = string;

export interface LimitOrderData {
    maker: Address,
    receiver: Address,
    makerAsset: Address,
    takerAsset: Address,
    makingAmount: string,
    takingAmount: string,
    makerTraits: bigint | string,
}

export type ExtensionParamsWithCustomData = ExtensionParams & {
    customData: string;
}

export interface ExtensionParams {
    makerAssetSuffix: string;
    takerAssetSuffix: string;
    makingAmountGetter: string;
    takingAmountGetter: string;
    predicate: string;
    permit: string;
    preInteraction: string;
    postInteraction: string;
}


export interface RFQOrderData {
    // Index number of RFQ limit order. Example: 1
    id: number;
    wrapEth?: boolean;
    // Timestamp when the RFQ limit order will expire (seconds). Example: 1623166102
    expiresInTimestamp: number;
    makerAssetAddress: string;
    takerAssetAddress: string;
    makingAmount: string;
    takingAmount: string;
    makerAddress: string;
    /**
     * Formerly takerAddress
     */
    allowedSender?: string; // Optional, by default = ZERO_ADDRESS
}

export type Address = string;

export type MakerTraits = string;

/**
 * Compatible with EIP712Object
 */
export type LimitOrder = {
    salt: string;
    maker: Address; // maker address
    receiver: Address;
    makerAsset: Address; // maker asset address
    takerAsset: Address; // taker asset address
    makingAmount: string;
    takingAmount: string;
    makerTraits: MakerTraits;
}

export type LimitOrderWithExtension = {
    order: LimitOrder;
    extension: string;
}

/**
 * Partial from LimitOrder
 */
export type LimitOrderInteractions = {
    offsets: bigint;
    interactions: string;
}

/**
 * uint40
 */
export type Nonce = number | bigint;

/**
 * seconds unit40
 */
export type PredicateTimestamp = number | bigint;

export const InteractionsFields = {
    makerAssetData: 0,
    takerAssetData: 1,
    getMakingAmount: 2,
    getTakingAmount: 3,
    predicate: 4,
    permit: 5,
    preInteraction: 6,
    postInteraction: 7,
// cuz enum has numeric keys also
} as const;

export type InteractionName = keyof typeof InteractionsFields;

export type Interactions = {
    [key in InteractionName]: string;
};

export interface RFQOrder {
    info: RFQOrderInfo;
    makerAsset: string;
    takerAsset: string;
    maker: string;
    allowedSender: string;
    makingAmount: string;
    takingAmount: string;
}

export enum LimitOrderProtocolMethods {
    getMakingAmount = 'getMakingAmount',
    getTakingAmount = 'getTakingAmount',
    arbitraryStaticCall = 'arbitraryStaticCall',
    fillOrder = 'fillOrder',
    fillOrderToWithPermit = 'fillOrderToWithPermit',
    fillOrderRFQ = 'fillOrderRFQ',
    cancelOrder = 'cancelOrder',
    cancelOrderRFQ = 'cancelOrderRFQ',
    nonce = 'nonce',
    advanceNonce = 'advanceNonce',
    increaseNonce = 'increaseNonce',
    and = 'and',
    or = 'or',
    eq = 'eq',
    lt = 'lt',
    gt = 'gt',
    timestampBelow = 'timestampBelow',
    timestampBelowAndNonceEquals = 'timestampBelowAndNonceEquals',
    nonceEquals = 'nonceEquals',
    remaining = 'remaining',
    transferFrom = 'transferFrom',
    checkPredicate = 'checkPredicate',
    remainingsRaw = 'remainingsRaw',
    simulate = 'simulate',
}
