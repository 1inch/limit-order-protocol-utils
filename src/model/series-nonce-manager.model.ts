/**
 * 0-255 uint8
 */
export type Series = bigint;

export const LimitOrderSeries = {
    /**
     * @deprecated Don't use!
     * Preserved just to indicate which values are already takken.
     */
    P2Pv2: <Series>BigInt(1),

    P2Pv3: <Series>BigInt(2),
    // TODO use for gasless bulk cancellation
    // Gaslessv3 = 3,
} as const;

export enum SeriesNonceManagerMethods {
    nonce = 'nonce',
    advanceNonce = 'advanceNonce',
    increaseNonce = 'increaseNonce',
    nonceEquals = 'nonceEquals',
}