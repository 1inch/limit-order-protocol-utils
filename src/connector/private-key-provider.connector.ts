import {ProviderConnector} from './provider.connector';
import Web3 from 'web3';
import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem} from '../model/abi.model';
import {AbiItem as Web3AbiItem} from 'web3-utils';
import {MsgParams, signTypedData_v4} from 'eth-sig-util';

export class PrivateKeyProviderConnector implements ProviderConnector {
    constructor(
        private readonly privateKey: string,
        protected readonly web3Provider: Web3
    ) {}

    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string {
        const contract = new this.web3Provider.eth.Contract(
            abi as Web3AbiItem[],
            address === null ? undefined : address
        );

        return contract.methods[methodName](...methodParams).encodeABI();
    }

    signTypedData(
        _walletAddress: string,
        typedData: EIP712TypedData,
        /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
        _typedDataHash = ''
    ): Promise<string> {
        const result = signTypedData_v4(Buffer.from(this.privateKey, 'hex'), {
            data: typedData,
            /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        } as MsgParams<any>);

        return Promise.resolve(result);
    }

    ethCall(contractAddress: string, callData: string): Promise<string> {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }

    decodeABIParameter<T>(type: string, hex: string): T {
        return this.web3Provider.eth.abi.decodeParameter(type, hex) as T;
    }
}
