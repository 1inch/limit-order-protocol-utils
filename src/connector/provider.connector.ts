import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem, AbiOutput} from '../model/abi.model';

export interface AbiDecodeResult {
    [key: string]: unknown,
}

export interface ProviderConnector {
    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string;

    signTypedData(
        walletAddress: string,
        typedData: EIP712TypedData,
        typedDataHash: string
    ): Promise<string>;

    ethCall(contractAddress: string, callData: string): Promise<string>;

    decodeABIParameter<T>(type: AbiOutput | string, hex: string): T;

    decodeABICallParameters(types: Array<AbiOutput | string>, callData: string): AbiDecodeResult;
}
