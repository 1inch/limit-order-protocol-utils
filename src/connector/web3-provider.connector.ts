import {ProviderConnector} from './provider.connector';
import Web3 from 'web3';
import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem} from '../model/abi.model';
import {AbiItem as Web3AbiItem} from 'web3-utils';

export class Web3ProviderConnector implements ProviderConnector {
    constructor(protected readonly web3Provider: Web3) {}

    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string {
        const contract = new this.web3Provider.eth.Contract(
            abi as Web3AbiItem[],
            address === null ? undefined : address
        );

        return contract.methods[methodName](...methodParams).encodeABI();
    }

    signTypedData(
        walletAddress: string,
        typedData: EIP712TypedData,
        /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
        _typedDataHash: string
    ): Promise<string> {
        if (!this.web3Provider.currentProvider) {
            throw new Error('Web3 currentProvider is null');
        }

        return this.web3Provider.currentProvider.send('eth_signTypedData_v4', [
            walletAddress,
            JSON.stringify(typedData),
        ]);
    }

    ethCall(contractAddress: string, callData: string): Promise<string> {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }

    decodeABIParameter<T>(type: string, hex: string): T {
        return this.web3Provider.eth.abi.decodeParameter(type, hex) as T;
    }
}
