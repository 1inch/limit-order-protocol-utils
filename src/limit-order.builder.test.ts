import {LimitOrderBuilder} from "./limit-order.builder";
import {ProviderConnector} from "./connector/provider.connector";
import {instance, mock, when} from "ts-mockito";
import {EIP712TypedData} from "./model/eip712.model";
import {EIP712_DOMAIN, ORDER_STRUCTURE, PROTOCOL_NAME, PROTOCOL_VERSION} from "./limit-order-protocol.const";

describe('LimitOrderBuilder - for build new limit order', () => {
    const contractAddress = '0xaaaaa';
    const chainId = 56;

    let providerConnector: ProviderConnector;
    let limitOrderBuilder: LimitOrderBuilder;

    beforeEach(() => {
        providerConnector = mock<ProviderConnector>();
        limitOrderBuilder = new LimitOrderBuilder(
            contractAddress,
            chainId,
            instance(providerConnector)
        );
    });

    it('buildOrderSignature() must call provider signTypedData', async () => {
        const walletAddress = '0xggggg';
        const typedData: EIP712TypedData = {
            primaryType: 'Order',
            types: {
                EIP712Domain: EIP712_DOMAIN,
                Order: ORDER_STRUCTURE
            },
            domain: {
                name: PROTOCOL_NAME,
                version: PROTOCOL_VERSION,
                chainId: chainId,
                verifyingContract: contractAddress
            },
            message: {
                foo: 'bar'
            },
        };
        const expectedSignature = '0x6767676767';

        when(providerConnector.signTypedData(walletAddress, typedData))
            .thenReturn(Promise.resolve(expectedSignature));

        const signature = await limitOrderBuilder.buildOrderSignature(walletAddress, typedData);

        expect(signature).toBe(expectedSignature);
    });
});
