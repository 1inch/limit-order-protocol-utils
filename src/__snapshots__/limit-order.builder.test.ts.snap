// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LimitOrderBuilder - for build new limit order Normal limit order buildLimitOrder() must create a limit order instance according to the given parameters 1`] = `
Object {
  "allowedSender": "0x0000000000000000000000000000000000000000",
  "getMakerAmount": "0xf4a215c300000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000064",
  "getTakerAmount": "0x296637bf00000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000064",
  "interaction": "0x333",
  "maker": "0xdddd91605c18a9999c1d47abfeed5daaaa700000",
  "makerAsset": "0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c",
  "makerAssetData": "0x23b872dd000000000000000000000000dddd91605c18a9999c1d47abfeed5daaaa70000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003",
  "makingAmount": "3",
  "permit": "0x222",
  "predicate": "0x111",
  "receiver": "0x0000000000000000000000000000000000000000",
  "salt": "1",
  "takerAsset": "0x111111111117dc0aa78b770fa6a738034120c302",
  "takerAssetData": "0x23b872dd0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000dddd91605c18a9999c1d47abfeed5daaaa7000000000000000000000000000000000000000000000000000000000000000000064",
  "takingAmount": "100",
}
`;

exports[`LimitOrderBuilder - for build new limit order Normal limit order buildLimitOrderHash() must create a hash of order with 0x prefix 1`] = `"0x775e4598649f19fdbfc36471af359d3c674dd269ca9e657596b6caf401276d8b"`;

exports[`LimitOrderBuilder - for build new limit order Normal limit order buildOrderSignature() must call the provider signTypedData method 1`] = `"0xf1a27c2b4ff94f6ea21542758b0f84b58bad811732a65abcefcfaaaff19477805ffe804686099ff30cb35351b5b01c2a66ec9fa1840ace5b5e08c58e4a552eef1b"`;

exports[`LimitOrderBuilder - for build new limit order RFQ limit order buildLimitOrder() must create an RFQ limit order instance according to the given parameters 1`] = `
Object {
  "allowedSender": "0x0000000000000000000000000000000000000000",
  "info": "29942129672714733576710520833",
  "maker": "0xdddd91605c18a9999c1d47abfeed5daaaa800000",
  "makerAsset": "0x111111111117dc0aa78b770fa6a738034120c302",
  "makingAmount": "5",
  "takerAsset": "0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c",
  "takingAmount": "600",
}
`;

exports[`LimitOrderBuilder - for build new limit order RFQ limit order buildLimitOrderHash() must create a hash of order with 0x prefix 1`] = `"0xb20ba7d659e73145309495c00cf664aaefd631e348254e9bdb7956130d1821a6"`;

exports[`LimitOrderBuilder - for build new limit order RFQ limit order buildOrderSignature() must call the provider signTypedData method 1`] = `"0x04add184953b8ef9bff07dd85c9682bc0c2ca33f45a78587e111aa99641c6d7c6b1d79e27670ca463c7da8f4a8e9553609294f5eeb535febde26f1ba4dd9298b1c"`;
